src/content/editor.ts

```diff
@@ -5,14 +5,62 @@ import { shouldSkipEditorDomChanges } from "./utilities";
 export const editorState = (port: Port, clickableCount: number) => {
   const source = activeElementSource();
   const cursor = activeElementCursor();
+  const filename = activeElementFilename();
 
-  port.postMessage({ source, cursor, clickableCount });
+  port.postMessage({ source, cursor, filename, clickableCount });
+};
+
+const activeElementIsCodeMirror = () => {
+  const codeMirrorNodes = document.querySelectorAll(".CodeMirror");
+  let isCodeMirror = false;
+  codeMirrorNodes.forEach((node) => {
+    if (node.contains(document.activeElement)) {
+      isCodeMirror = true;
+    }
+  });
+  return isCodeMirror;
+};
+
+const activeElementIsMonaco = () => {
+  const monacoNodes = document.querySelectorAll(".monaco-editor");
+  let isMonaco = false;
+  monacoNodes.forEach((node) => {
+    if (node.contains(document.activeElement)) {
+      isMonaco = true;
+    }
+  });
+  return isMonaco;
 };
 
 // Finds the active element and gets its user-visible source in plaintext.
 const activeElementSource = () => {
   let activeElementSource: string | null = null;
-  if (document.activeElement) {
+  document.addEventListener(
+    "serenade-chrome-send-codemirror",
+    (e) => {
+      if ((e as any).detail.codeMirrorValue) {
+        activeElementSource = (e as any).detail.codeMirrorValue;
+      }
+    },
+    {
+      once: true,
+    }
+  );
+  document.dispatchEvent(new CustomEvent("serenade-chrome-request-codemirror"));
+  document.addEventListener(
```

Comment:
I think there might be a race condition here, where we send the `serenade-chrome-request-codemirror` event but don't get a response back before the rest of the code runs and we get to `return activeElementSource;` before it's set here. Maybe it would be safest to wrap the three cases into branches with `if activeElementIsCodeMirror()` etc?

Reply:
Good catch. Added in fbce206

Reply:
Sorry, I think the race condition could actually still exist because the event listener doesn't necessarily get called before the return, even if we do branch. I think there's something you could do with promises, but if this is working well for you every time, it might not be an issue worth trying to fix.